### DISCUSSION TITLE
Create alternative to IXunitSerializable pattern?

### DISCUSSION THREAD

**Discussion by lemonlion** [xunit/xunit#2570] [Ideas]
From what @bradwilson has said in previous comments/discussions (eg https://github.com/xunit/xunit/issues/1679), for MemberData/ClassData, the only recommended way of getting the test explorer to run tests rows completely independently rather than together is to have the objects under test inherit from IXunitSerializable.   The reason is unpredictable behaviour in classes/records in situations where the implementation of the properties gives non repeatable results when serializing.  eg

```
public class InconsistentlySerializableClass
{
    public int ProblematicProperty => new Random().Next();
}
```

This is fair enough, however I feel it would be useful if the user was able to mark objects they're using as serializable explicitly when they know they are consistently serializable, and do so by mechanisms that are more flexible than inheriting from IXunitSerializable.  For example:

For this given record:

```
record Foo(string Bar1, string Bar2);
```

It would be good to be able to implement one of these patterns:

```
[Theory]
[MemberData(nameof(GetData))]
public void MyTestUsingASerializableInlineAttribute([Serializable] Foo a, [Serializable] Foo b)
{ .... }
```
or

```
[Theory(AllInputsSerializable = true)]
[MemberData(nameof(GetData))]
public void MyTestUsingAFlagInTheoryConstructor(Foo a, Foo b)
{ .... }
```

Either of the above options above gives the user far more flexibility, while still making them assert explicitly that there is no issue reliably serializing their objects in an automated manner.  This is far preferable to having to make Foo implement IXunitSerializable, as that solution is not practical for pre-existing objects, and highly cumbersome/verbose in the example above where you're declaring a simple one line ordered record.

Is this something we could consider in future versions?

**Comment by bradwilson** [comment]
What problem are you trying to solve? You don't want to implement `IXunitSerializable`?

**Reply by lemonlion** [reply]
Yes, I'd like to not have to implement `IXunitSerializable` in situations where I declare that everything passed in is safe for automatic serialization.

I'd like to be able to take pre-existing classes or records, and just tell the test that for this test theory, and this test theory alone, I can guarantee that all the objects it's getting as constructor parameters are fully automatically serializable.  I'm suggesting we tell the test that by either specifying it in the Theory constructor, or by marking the individual objects with inline 'Serializable' attributes.  

It is so easy to make new ordered records now to use as objects for tests, so having to implement serialization for them is a pain when you know their properties are all just simple strings and ints.  Also, I've had cases where the method under test already takes a set complex (but fully serializable) object, so it's much easier to feed that in to the test itself, and I don't want to have to inherit that object and then implement IXUnitSerializable and custom serialization for it before passing it in to the test.

Again I completely understand why this isn't the `default` behaviour, but I think it's fine when the user has explicitly declared that these objects are safe for automatic serialization.

**Reply by bradwilson** [reply]
Generalized object serialization was removed for security concerns (all the potential built-in ways to do so have been deprecated for that reason). https://github.com/xunit/xunit/discussions/2506

`IXunitSerializable` is here to stay.

**Reply by lemonlion** [reply]
> `IXunitSerializable` is here to stay.

A surprising response - I didn't advocate removing it.

I was asking for additional mechanisms to achieve general serialization for specific use cases through explicit flags in theory constructors or inline attributes.  As far as I can see they could still be achieved by using a fully secure serialization mechanism like System.Text.Json.  I understand that that would affect polymorphism, but again, I'm not talking about default behaviour, I'm talking about an optional flag that you can set in a TheoryAttribute constructor (or, again, inline attributes).  Cases where polymorphism is necessary do not necessarily always cross over with these use cases, therefore giving a user the ability to turn this functionality on and off on a per test case basis would benefit them greatly.  As XUnit is basically the _defacto_ testing framework in .Net I think it would be good to take a non prescriptive approach here and give the users as many options as possible.
